### ISSUE TITLE
Investigate module systems

### ISSUE THREAD

**Issue by arrdem** [jaunt-lang/jaunt#58]
Related to #5, #6 

The next logical step is to have a clear, statically defined way to reason about the exports of a namespace as well as its imports. Clojure exports everything by default, which is alright, but makes analysis without code loading impossible. Maybe a good idea, maybe not, certainly a change from the way that Clojure operates.

**Comment by yurrriq** [comment]
:+1: for explicit exports. Erlang's behaviour, for example, is the opposite, i.e. functions are _not_ exported by default, and must be explicitly exported.

**Comment by arrdem** [comment]
Yeah Haskell does this as well if I recall, as do most languages with a module system. The idea that sparked this is wanting to do some sort of package manifest in terms of exported symbols and their arities/types and then seriously reason about package/dependency updates ala Elm.

**Comment by yurrriq** [comment]
I'm pretty sure you're right about Haskell. Oh man. That sounds exactly like [LFE](https://github.com/rvirding/lfe). My [LFE port of Codox](https://github.com/quasiquoting/lodox) would've been wayy more difficult if [`Mod:module_info/{0,1}`](http://erlang.org/doc/reference_manual/modules.html#id77156)) didn't exist..
